File: scalable-rest-api-solution-pro/src/app/controllers/ProjectController.cpp
```cpp
#include "ProjectController.h"
#include "core/utils/Logger.h"
#include "core/utils/Config.h" // For getting user_id from request attributes
#include "app/services/ProjectService.h"
#include <stdexcept>
#include <string>

using namespace Pistache;
using namespace nlohmann;

void ProjectController::getProjects(const Rest::Request& request, Http::ResponseWriter response) {
    try {
        long long user_id = request.getAttribute<long long>("user_id").get(); // Extracted by AuthMiddleware
        std::vector<Project> projects = ProjectService::getAllProjects(user_id);
        json projects_json = json::array();
        for (const auto& project : projects) {
            projects_json.push_back(project.toJson());
        }
        response.send(Http::Code::Ok, projects_json.dump(), MIME(Application, Json));
    } catch (const std::exception& e) {
        Logger::error("ProjectController::getProjects - Error: {}", e.what());
        response.send(Http::Code::Internal_Server_Error, json({{"message", e.what()}}).dump());
    }
}

void ProjectController::getProjectById(const Rest::Request& request, Http::ResponseWriter response) {
    try {
        long long project_id = request.param(":id").as<long long>();
        long long user_id = request.getAttribute<long long>("user_id").get();

        std::optional<Project> project_opt = ProjectService::getProjectById(project_id, user_id);
        if (project_opt) {
            response.send(Http::Code::Ok, project_opt->toJson().dump(), MIME(Application, Json));
        } else {
            response.send(Http::Code::Not_Found, json({{"message", "Project not found or you don't have access"}}).dump());
        }
    } catch (const std::exception& e) {
        Logger::error("ProjectController::getProjectById - Error: {}", e.what());
        response.send(Http::Code::Internal_Server_Error, json({{"message", e.what()}}).dump());
    }
}

void ProjectController::createProject(const Rest::Request& request, Http::ResponseWriter response) {
    try {
        json request_body = json::parse(request.body());
        Project new_project = Project::fromJson(request_body);
        
        if (new_project.name.empty() || new_project.description.empty()) {
            response.send(Http::Code::Bad_Request, json({{"message", "Project name and description are required"}}).dump());
            return;
        }

        long long user_id = request.getAttribute<long long>("user_id").get();
        new_project = ProjectService::createProject(new_project, user_id);
        response.send(Http::Code::Created, new_project.toJson().dump(), MIME(Application, Json));
    } catch (const json::parse_error& e) {
        Logger::warn("ProjectController::createProject - Invalid JSON: {}", e.what());
        response.send(Http::Code::Bad_Request, json({{"message", "Invalid JSON body"}}).dump());
    } catch (const std::out_of_range& e) {
        Logger::warn("ProjectController::createProject - Missing JSON field: {}", e.what());
        response.send(Http::Code::Bad_Request, json({{"message", "Missing required JSON fields (name, description)"}}).dump());
    } catch (const std::exception& e) {
        Logger::error("ProjectController::createProject - Error: {}", e.what());
        response.send(Http::Code::Internal_Server_Error, json({{"message", e.what()}}).dump());
    }
}

void ProjectController::updateProject(const Rest::Request& request, Http::ResponseWriter response) {
    try {
        long long project_id = request.param(":id").as<long long>();
        long long user_id = request.getAttribute<long long>("user_id").get();
        json request_body = json::parse(request.body());
        
        Project updated_project_data = Project::fromJson(request_body);
        
        std::optional<Project> updated_project_opt = ProjectService::updateProject(project_id, updated_project_data, user_id);
        if (updated_project_opt) {
            response.send(Http::Code::Ok, updated_project_opt->toJson().dump(), MIME(Application, Json));
        } else {
            response.send(Http::Code::Not_Found, json({{"message", "Project not found or you don't have access to update"}}).dump());
        }

    } catch (const json::parse_error& e) {
        Logger::warn("ProjectController::updateProject - Invalid JSON: {}", e.what());
        response.send(Http::Code::Bad_Request, json({{"message", "Invalid JSON body"}}).dump());
    } catch (const std::exception& e) {
        Logger::error("ProjectController::updateProject - Error: {}", e.what());
        response.send(Http::Code::Internal_Server_Error, json({{"message", e.what()}}).dump());
    }
}

void ProjectController::deleteProject(const Rest::Request& request, Http::ResponseWriter response) {
    try {
        long long project_id = request.param(":id").as<long long>();
        long long user_id = request.getAttribute<long long>("user_id").get();

        if (ProjectService::deleteProject(project_id, user_id)) {
            response.send(Http::Code::No_Content); // 204 No Content for successful deletion
        } else {
            response.send(Http::Code::Not_Found, json({{"message", "Project not found or you don't have access to delete"}}).dump());
        }
    } catch (const std::exception& e) {
        Logger::error("ProjectController::deleteProject - Error: {}", e.what());
        response.send(Http::Code::Internal_Server_Error, json({{"message", e.what()}}).dump());
    }
}
```

File: production-rest-api-backend-advanced/pms-api/backend/src/modules/projects/project.controller.ts
import { Request, Response, NextFunction } from 'express';
import { ProjectService } from './project.service';
import { CreateProjectDto, UpdateProjectDto } from './project.dtos';
import { ApiError } from '../../utils/apiError';

export class ProjectController {
  private projectService = new ProjectService();

  getAllProjects = async (req: Request, res: Response, next: NextFunction) => {
    try {
      const projects = await this.projectService.getAllProjects();
      res.status(200).json(projects);
    } catch (error) {
      next(error);
    }
  };

  getProjectById = async (req: Request, res: Response, next: NextFunction) => {
    try {
      const project = await this.projectService.getProjectById(req.params.id);
      res.status(200).json(project);
    } catch (error) {
      next(error);
    }
  };

  createProject = async (req: Request, res: Response, next: NextFunction) => {
    try {
      const createDto: CreateProjectDto = req.body;
      if (!createDto.name) {
        throw new ApiError(400, 'Project name is required.');
      }
      if (!req.user?.id) {
        throw new ApiError(401, 'User not authenticated to create project.');
      }
      const newProject = await this.projectService.createProject(createDto, req.user.id);
      res.status(201).json(newProject);
    } catch (error) {
      next(error);
    }
  };

  updateProject = async (req: Request, res: Response, next: NextFunction) => {
    try {
      const updateDto: UpdateProjectDto = req.body;
      if (Object.keys(updateDto).length === 0) {
        throw new ApiError(400, 'At least one field (name, description) must be provided for update.');
      }
      if (!req.user?.id) {
        throw new ApiError(401, 'User not authenticated to update project.');
      }
      const updatedProject = await this.projectService.updateProject(req.params.id, updateDto, req.user.id);
      res.status(200).json(updatedProject);
    } catch (error) {
      next(error);
    }
  };

  deleteProject = async (req: Request, res: Response, next: NextFunction) => {
    try {
      if (!req.user?.id) {
        throw new ApiError(401, 'User not authenticated to delete project.');
      }
      await this.projectService.deleteProject(req.params.id, req.user.id);
      res.status(204).send(); // No content for successful deletion
    } catch (error) {
      next(error);
    }
  };
}